### PR TITLE
Add preview to XfD module for afd, mfd, tfd/tfm, ffd

### PR DESCRIPTION
**Pull request by ndrewh**
It might work for other venues too but haven't tested.
Issue #168

This is a GCI task.

### CODE REVIEW

**Comment by atlight** [comment]
My main comment is that I would prefer to have one central function that determines the wikitext for a particular deletion nomination. The way you have it in this patch set, the wikitext for preview is generated in one function, and the actual wikitext for the nomination is generated elsewhere. Having duplicated logic like this is wasteful, and the chance of them getting out of sync is high.

I'd like to see a unified approach. For example, you could implement a function `Twinkle.xfd.callbacks.afd.getWikitext` which takes a `params` object and returns wikitext. This would then be called both by the previewing logic and the actual editing logic.

File: modules/twinklexfd.js [hunk 1]
```diff
@@ -569,8 +645,7 @@ Twinkle.xfd.callbacks = {
 		discussionPage: function(pageobj) {
 			var params = pageobj.getCallbackParameters();
 
-			pageobj.setPageText("{{subst:afd2|text=" + Morebits.string.formatReasonText(params.reason) +
-				" ~~~~|pg=" + Morebits.pageNameNorm + "|cat=" + params.xfdcat + "}}\n");
+			pageobj.setPageText(Twinkle.xfd.callbacks.getDiscussionWikitext("afd", params));
```
Comment (atlight) [inline modules/twinklexfd.js]: Yeah! I like it!

**Comment by atlight** [comment]
Good, just one comment: The preview box should be closed whenever the top venue dropdown is changed.

**Comment by ndrewh** [comment]
Ok, I'll fix that later today

**Comment by ndrewh** [comment]
Fixed!

File: modules/twinklexfd.js [hunk 2]
```diff
@@ -700,19 +777,7 @@ Twinkle.xfd.callbacks = {
...
-				break;
-			case 'tfm':
-				added_data = "{{subst:tfm2|text=" + Morebits.string.formatReasonText(params.reason) +
-					" ~~~~|1=" + mw.config.get('wgTitle') + "|2=" + params.target + "}}";
-				break;
-			default:
-				alert("twinklexfd in todaysList: unknown TFD action");
-				break;
-			}
+			added_data = Twinkle.xfd.callbacks.getDiscussionWikitext(params.xfdcat, params);
```
Comment (mc10) [inline modules/twinklexfd.js]: Minor: You can just merge this with the previous line.

File: modules/twinklexfd.js [hunk 3]
```diff
@@ -442,6 +455,71 @@ Twinkle.xfd.callback.change_category = function twinklexfdCallbackChangeCategory
...
+
+		if (params.target) {
+			text += "|2=" + params.target;
+		}
+
+		if (params.uploader) {
+			text += "|Uploader=" + params.uploader;
+		}
+
+		text += "}}"
```
Comment (mc10) [inline modules/twinklexfd.js]: Minor: Missing a semi.

File: modules/twinklexfd.js [hunk 4]
```diff
@@ -442,6 +455,71 @@ Twinkle.xfd.callback.change_category = function twinklexfdCallbackChangeCategory
...
+		if (form.xfdtarget) {
+			params.target = form.xfdtarget.value;
+		}
+		if (venue === "ffd") {
+			// Fetch the uploader
+			var page = new Morebits.wiki.page(mw.config.get('wgPageName'));
+			page.lookupCreator(function() {
+				params.uploader = page.getCreator();
+				Twinkle.xfd.callbacks.showPreview(form, venue, params);
+			})
```
Comment (mc10) [inline modules/twinklexfd.js]: Minor: Missing a semi.

File: modules/twinklexfd.js (hunk 2 above)
Comment (ndrewh) [inline modules/twinklexfd.js]: done

File: modules/twinklexfd.js (hunk 3 above)
Comment (ndrewh) [inline modules/twinklexfd.js]: done

File: modules/twinklexfd.js (hunk 4 above)
Comment (ndrewh) [inline modules/twinklexfd.js]: done

**Comment by ndrewh** [comment]
Fixed again!

**Comment by atlight** [comment]
Great work as always!

**Pull request merged**